### PR TITLE
Fix Wallet creation issues

### PR DESCRIPTION
**Pull request by simoarpe**
## Resolves https://github.com/brave/brave-browser/issues/39608

Close web Wallet and force new tab upon creation.
Slack discussion: https://bravesoftware.slack.com/archives/C04JJNFRPK7/p1721741104721669

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [X] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1. Create a new Wallet and follow the onboarding flow to completion. Recovery steps and fingerprint unlock can be skipped.
2. Go to settings > Wallet > Reset and clear wallet data > Confirm typing "Yes".
3. Create again a new Wallet and follow the onboarding flow to completion. Recovery steps and fingerprint unlock can be skipped.

Repeat the steps 1, 2 and 3 and observe that the Wallet page in `brave://wallet/crypto/portfolio/assets` is always loaded correctly.

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 1]
```diff
@@ -1784,6 +1787,48 @@ public Profile getCurrentProfile() {
         return Profile.fromWebContents(tab.getWebContents());
     }
 
+    public void closeExistingTabs(final boolean closeActiveTab, @NonNull final String baseUrl) {
+        final Tab activeTab = getActivityTab();
+        final TabModel tabModel = getCurrentTabModel();
+        if (activeTab != null
+                && activeTab.getUrl().getSpec().startsWith(baseUrl)
```
Comment (supermassive) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:1794]: comparing urls with startWith does not always work correct
like `brave://wallet` and `brave://walletsomething` would be a false positive
Most likely it should be implemented as searching for tabs by origin equality

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 2]
```diff
@@ -1784,6 +1787,48 @@ public Profile getCurrentProfile() {
...
+            if (tab != null) {
+                tab.setClosing(true);
+                tabModel.closeTab(tab);
+            }
+        }
+    }
+
+    public Tab selectExistingBaseUrlTab(@NonNull final String baseUrl) {
+        Tab tab = getActivityTab();
+        if (tab != null && tab.getUrl().getSpec().startsWith(baseUrl)) {
```
Comment (supermassive) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:1813]: same issue with `startsWith`

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/WalletUtils.java [hunk 3]
```diff
@@ -106,16 +107,31 @@ public static AccountInfo getAccountInfoFromIntent(@NonNull final Intent intent)
...
     }
 
-    public static void openWebWallet() {
+    public static void openWebWallet(final boolean forceNewTab) {
         try {
             BraveActivity activity = BraveActivity.getBraveActivity();
-            activity.openNewOrSelectExistingTab(BraveActivity.BRAVE_WALLET_URL, true);
+            if (forceNewTab) {
+                activity.openAlwaysNewAndCloseExistingTab(
+                        BraveActivity.BRAVE_WALLET_BASE_URL, BraveActivity.BRAVE_WALLET_URL);
```
Comment (supermassive) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/WalletUtils.java:115]: I would suggest to implement that as  `if (forceNewTab) {closeAllTabsByOrigin(...)`} followed by `openNewOrRefreshExistingTab(...)`

**Review by supermassive — APPROVED** [review]
lgtm with suggestions
I think someone with android-specific knowledge should also take a look

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java (hunk 1 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/app/BraveActivity.java]: 👍

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java (hunk 2 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/app/BraveActivity.java:1812]: 👍

**Comment by simoarpe** [comment]
Thanks for the precious review @supermassive ! I'm asking also @SergeyZhukovsky to review this.

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 4]
```diff
@@ -1784,6 +1787,46 @@ public Profile getCurrentProfile() {
...
+    /** Close all tabs whose URL origin matches with a given origin. */
+    public void closeAllTabsByOrigin(@NonNull final String origin) {
+        final Tab activeTab = getActivityTab();
+        final TabModel tabModel = getCurrentTabModel();
+        if (activeTab != null && activeTab.getUrl().getOrigin().getSpec().equals(origin)) {
+            activeTab.setClosing(true);
+            tabModel.closeTab(activeTab);
+        }
+
+        Set<Integer> tabIndexes = TabModelUtils.getTabIndexesByUrlOrigin(tabModel, origin);
```
Comment (simoarpe) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:1799]: Sadly, this class is not part of our codebase @supermassive. Any workaround to suggest?

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 5]
```diff
@@ -1784,6 +1789,68 @@ public Profile getCurrentProfile() {
         return Profile.fromWebContents(tab.getWebContents());
     }
 
+    /** Close all tabs whose URL origin matches with a given origin. */
+    public void closeAllTabsByOrigin(@NonNull final String origin) {
+        final Tab activeTab = getActivityTab();
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:1794]: Doesn't current Activity tab included in `Set<Integer> tabIndexes = getTabIndexesByUrlOrigin(tabModel, origin);`? I think it should return it's index there as well.

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 6]
```diff
@@ -1784,6 +1789,68 @@ public Profile getCurrentProfile() {
...
+                tab.setClosing(true);
+                tabModel.closeTab(tab);
+            }
+        }
+    }
+
+    public Tab selectExistingUrlOriginTab(@NonNull final String origin) {
+        Tab tab = getActivityTab();
+        if (tab != null && tab.getUrl().getOrigin().getSpec().equals(origin)) {
+            return tab;
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:1814]: the same questions about the tab indexes Set

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java (hunk 5 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/app/BraveActivity.java:1794]: Yes, you're right! Good catch! I'm going to refactor this part.

Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/app/BraveActivity.java]: Changes done in commit [`91727f8` (#24960)](https://github.com/brave/brave-core/pull/24960/commits/91727f880a0647d3271072b698cfd760dddad24d)

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java (hunk 6 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/app/BraveActivity.java]: Changes done in commit [`91727f8` (#24960)](https://github.com/brave/brave-core/pull/24960/commits/91727f880a0647d3271072b698cfd760dddad24d)

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/settings/BraveWalletResetPreference.java [hunk 7]
```diff
@@ -81,27 +82,27 @@ private void showBraveWalletResetDialog() {
...
-
-                                for (String key : WalletConstants.BRAVE_WALLET_PREFS) {
-                                    ChromeSharedPreferences.getInstance().removeKey(key);
-                                }
+                (dialog, button) -> {
+                    if (button == AlertDialog.BUTTON_POSITIVE) {
+                        String inputText = input.getText().toString().trim();
+                        if (TextUtils.equals(inputText, mConfirmationPhrase)) {
+                            Log.w(TAG, "Reset");
+                            WalletNativeUtils.resetWallet(Utils.getProfile(false));
```
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/settings/BraveWalletResetPreference.java:90]: Do not use static methods like this to get the profile. See `SignInPreference` in upstream

Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/settings/BraveWalletResetPreference.java:90]: yeah, we can definitely do how it is inside `SignInPreference` and `MainSettings`. The main settings received `getProfile()` call recently, it didn't have it before, but right now we should use it.

Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/settings/BraveWalletResetPreference.java:90]: Becasue this PR has been already merged, I think it's best to track this kind of suggestions filing a new GitHub issue or fixing them directly when it's quick. So, here we go @bridiver, @SergeyZhukovsky: https://github.com/brave/brave-core/pull/28608

**Pull request merged**